### PR TITLE
Add tests for autodiff of operators

### PR DESCRIPTION
**Pull request by ckolbPTB**
Adds the gradient checks for operators from #307

### CODE REVIEW

**Comment by ckolbPTB** [comment]
The MOLLI test fails. Currently we have:
`signal(a,b,t1) = a * (1 - b/a * torch.exp(ti / t1 * (1 - b/a)))`

The b/a term is what is causing problems. My suggestion would be to change this to 
`signal(a,c,t1) = a * (1 - c * torch.exp(ti / t1 * (1 - c)))`

a and b don't have any physiological meaning and this should also make it easier to find good constraints for c. 

See #406

**Comment by ckolbPTB** [comment]
by the way I think it should be `ti / t1 * (b/a - 1)`

Messroghli et al:

grafik

File: tests/operators/models/test_inversion_recovery.py [hunk 1]
```diff
@@ -39,3 +40,11 @@ def test_inversion_recovery_shape(parameter_shape, contrast_dim_shape, signal_sh
     m0, t1 = create_parameter_tensor_tuples(parameter_shape, number_of_tensors=2)
     (signal,) = model_op.forward(m0, t1)
     assert signal.shape == signal_shape
+
+
+@pytest.mark.filterwarnings('ignore:Anomaly Detection has been enabled')
```
Comment (fzimmermann89) [inline tests/operators/models/test_inversion_recovery.py]: can we isntead filter the warnings inside the test helper?

[i.e. use  filterwarnings](https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings)

or ignore this warning globally in the tests (set it in the pyproject.toml)

File: tests/helper.py [hunk 2]
```diff
@@ -73,19 +79,19 @@ def dotproduct_adjointness_test(
     torch.testing.assert_close(dotproduct_range, dotproduct_domain, rtol=relative_tolerance, atol=absolute_tolerance)
 
 
-def operator_isometry_test(
-    operator: Operator, u: torch.Tensor, relative_tolerance: float = 1e-3, absolute_tolerance=1e-5
+def linear_operator_isometry_test(
+    linear_operator: LinearOperator, u: torch.Tensor, relative_tolerance: float = 1e-3, absolute_tolerance=1e-5
```
Comment (fzimmermann89) [inline tests/helper.py]: why is this limited to linear_operator?
this tests should work for  any operators first output

File: tests/helper.py [hunk 3]
```diff
@@ -26,9 +28,13 @@ def relative_image_difference(img1: torch.Tensor, img2: torch.Tensor) -> torch.T
 
 
 def dotproduct_adjointness_test(
-    operator: Operator, u: torch.Tensor, v: torch.Tensor, relative_tolerance: float = 1e-3, absolute_tolerance=1e-5
+    linear_operator: LinearOperator,
```
Comment (fzimmermann89) [inline tests/helper.py]: not to fond of this change.
that it is a linear_operator is already given by the type info.
does it ake the code easier to understand to read linear_operator than operator?
if talking about it, I would always say, for example: "Given a linear operator, you can always get the adjoint of the _operator_", i.e. drop the adjective.


but if you have strong preference for linear_operator here, I wont care.

File: tests/helper.py [hunk 4]
```diff
@@ -128,4 +134,37 @@ def operator_unitary_test(
     AssertionError
         if the adjointness property does not hold
     """
-    torch.testing.assert_close(u, operator.adjoint(operator(u)[0])[0], rtol=relative_tolerance, atol=absolute_tolerance)
+    torch.testing.assert_close(
+        u, linear_operator.adjoint(linear_operator(u)[0])[0], rtol=relative_tolerance, atol=absolute_tolerance
+    )
+
+
+def autodiff_of_operator_test(
```
Comment (fzimmermann89) [inline tests/helper.py]:
```suggestion
def autodiff_test(
```

File: tests/helper.py [hunk 5]
```diff
@@ -128,4 +134,37 @@ def operator_unitary_test(
...
+    operator: Operator[*tuple[torch.Tensor, ...], tuple[torch.Tensor, ...]],
+    *u: Any,
+):
+    """Test if autodiff of an operator is working.
+    This test does not check that the gradient is correct but simply that it can be calculated using autodiff.
```
Comment (fzimmermann89) [inline tests/helper.py]: ..using both torch.func.jvp and torch.func.vjp

**Comment by fzimmermann89** [comment]
remind me again, we we have no grad_check in the helpers, testing the numeric values of the gradients?

**Comment by ckolbPTB** [comment]
> remind me again, we we have no grad_check in the helpers, testing the numeric values of the gradients?

I don't think we have discussed this yet. If we use pytorch autograd functionality, would this test not merely test pytorch functionality?

File: pyproject.toml [hunk 6]
```diff
@@ -66,7 +66,8 @@ notebook = [
 testpaths = ["tests"]
 filterwarnings = [
     "error",
-    "ignore:'write_like_original':DeprecationWarning:pydicom:"
+    "ignore:'write_like_original':DeprecationWarning:pydicom:",
+    "ignore:Anomaly Detection has been enabled:UserWarning", #torch.autograd
```
Comment (ckolbPTB) [inline pyproject.toml]: I am a bit puzzled here because I thought that something like 
`ignore:Anomaly Detection has been enabled:UserWarning:torch`
or
`ignore:Anomaly Detection has been enabled:UserWarning:torch.*`
should work, but I could not get it to work. If anybody knows how to get it to work with specifying the module correctly I would be very happy to know.

**Comment by ckolbPTB** [comment]
@fzimmermann89 could you have a look at this again please

File: tests/operators/models/test_inversion_recovery.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 import pytest
 import torch
 from mrpro.operators.models import InversionRecovery
+from tests.helper import autodiff_test
```
Comment (fzimmermann89) [inline tests/operators/models/test_inversion_recovery.py]: we could consider importing all these in `tests/__init__.py`
Such that we can just do from tests import ..., similar as we do with the randomgenerator. 
I think we realistically only ever import from 'helper' of tests anyways...

**Review by fzimmermann89 — COMMENTED** [review]
remind me, what was the plan here?
should we add a test for each operator and its adjoint?

**Comment by fzimmermann89** [comment]
So far, what is there look fine to me

**Comment by ckolbPTB** [comment]
> remind me, what was the plan here?
> should we add a test for each operator and its adjoint?

With this PR we wanted to add autodiff tests for all non-linear operators. autodiff (and some more) tests for linear operators would then we added in PR #407

File: tests/helper.py [hunk 8]
```diff
@@ -129,3 +138,32 @@ def operator_unitary_test(
         if the adjointness property does not hold
     """
     torch.testing.assert_close(u, operator.adjoint(operator(u)[0])[0], rtol=relative_tolerance, atol=absolute_tolerance)
+
+
+def autodiff_test(
+    operator: Operator[*tuple[torch.Tensor, ...], tuple[torch.Tensor, ...]],
+    *u: Any,
```
Comment (fzimmermann89) [inline tests/helper.py]: why is this any?

Comment (fzimmermann89) [reply tests/helper.py]: shoulnt this match *tuple[torch.Tensor, ...] from operator?

Comment (fzimmermann89) [reply tests/helper.py]: i will try if that works

Comment (fzimmermann89) [reply tests/helper.py]: also, this is not py310 compatible, we would have to use Unpack instead of splitting with *

**Comment by fzimmermann89** [comment]
Will need #515 to pass mypy

**Comment by fzimmermann89** [comment]
> Will need #515 to pass mypy

I stand corrected -- #515 would make the type hints more useful, but it passes without the changes

**Review by fzimmermann89 — APPROVED** [review]
